File: decompile/ScriptImporter.js
const fs = require('fs');
const path = require('path');
const crc32 = require('crc').crc32;

function getAllFiles(dirPath)
{
    let fileList = [];
    const files = fs.readdirSync(dirPath);
    let directoryList = [];

    files.forEach((file) => 
    {
        const fullPath = path.join(dirPath, file);

        const stat = fs.statSync(fullPath);

        if (stat.isDirectory())
        {
            //fileList = fileList.concat(getAllFiles(fullPath));
            directoryList.push(fullPath);
        }
        else
        {
            fileList.push(fullPath);
        }
    });

    directoryList.forEach((directory) =>
    {
        fileList = fileList.concat(getAllFiles(directory));
    });

    return fileList;
}

function write7BitEncodedInt(bufferArray, value)
{
    while (value >= 0x80)
    {
        bufferArray.push(Buffer.from([value | 0x80]));
        value >>= 7;
    }
    bufferArray.push(Buffer.from([value]));
}

function writeBinaryString(bufferArray, str)
{
    const encoder = new TextEncoder();
    const encodedString = encoder.encode(str);
    write7BitEncodedInt(bufferArray, encodedString.length);
    bufferArray.push(Buffer.from(encodedString));
}

function createBinaryFile(inputDir, inputBinaryFile, outputBinaryFile)
{
    const dataBuffer = fs.readFileSync(inputBinaryFile);

    const bufferArray = [];
    var magic = dataBuffer.readInt32LE(0);
    var tmpVal = dataBuffer.readInt32LE(4);
    var version = dataBuffer.readInt32LE(8);

    const fileList = getAllFiles(inputDir);
    const headerBuffer = Buffer.alloc(12);
    headerBuffer.writeInt32LE(magic, 0);
    headerBuffer.writeInt32LE(tmpVal, 4);
    headerBuffer.writeInt32LE(version, 8);
    bufferArray.push(headerBuffer);

    const fileCountBuffer = Buffer.alloc(4);
    fileCountBuffer.writeInt32LE(fileList.length, 0);
    bufferArray.push(fileCountBuffer);
    var count = 0;

    fileList.forEach((filePath) =>
    {
        var fileName = path.relative(inputDir, filePath);
        fileName = fileName.replace(/\\/g, '/');
        const fileContent = fs.readFileSync(filePath);

        writeBinaryString(bufferArray, fileName);

        const contentLengthBuffer = Buffer.alloc(4);
        contentLengthBuffer.writeInt32LE(fileContent.length, 0);
        bufferArray.push(contentLengthBuffer);

        bufferArray.push(fileContent);

        console.log(`${count} pushed.`);
    });

    const finalBuffer = Buffer.concat(bufferArray);
    fs.writeFileSync(outputBinaryFile, finalBuffer);

    console.log(`Binary file created at: ${outputBinaryFile}`);
}

function createCrcSizeFile(outputBinaryFile, outputCrcFile)
{
    const dataBuffer = fs.readFileSync(outputBinaryFile);

    const length = dataBuffer.length;
    const bufferCrc32 = crc32(dataBuffer);

    const outputContent = `${length}|${bufferCrc32}`;
    fs.writeFileSync(outputCrcFile, outputContent);
    console.log(`Crc32 result exported to: ${outputCrcFile}`);
}

const inputDirectory = './export';
const inputBinaryFile = 'LWScripts.data';
const outputBinaryFile = 'LWScriptsNew.data';
const outputCrcFile = 'LWScriptsNew.txt';

if (!fs.existsSync(inputDirectory))
{
    console.error(`Input director "${inputDirectory}" does not exist.`);
    process.exit(1);
}

createBinaryFile(inputDirectory, inputBinaryFile, outputBinaryFile);
createCrcSizeFile(outputBinaryFile, outputCrcFile);

File: decompile/ScriptImporter.py
import os
import struct
import zlib
from io import BytesIO


def get_all_files(dir_path):
    file_list = []
    directory_list = []

    for root, dirs, files in os.walk(dir_path):
        for file in files:
            full_path = os.path.join(root, file)
            file_list.append(full_path)
        for dir in dirs:
            directory_list.append(os.path.join(root, dir))

    return file_list


def write_7bit_encoded_int(buffer_array, value):
    while value >= 0x80:
        buffer_array.append(bytes([value | 0x80]))
        value >>= 7
    buffer_array.append(bytes([value]))


def write_binary_string(buffer_array, string):
    encoded_string = string.encode('utf-8')
    write_7bit_encoded_int(buffer_array, len(encoded_string))
    buffer_array.append(encoded_string)


def create_binary_file(input_dir, input_binary_file, output_binary_file,progress_callback):
    with open(input_binary_file, 'rb') as file:
        data_buffer = file.read()

    buffer_array = []
    magic = struct.unpack('<I', data_buffer[0:4])[0]
    tmp_val = struct.unpack('<I', data_buffer[4:8])[0]
    version = struct.unpack('<I', data_buffer[8:12])[0]

    file_list = get_all_files(input_dir)

    header_buffer = struct.pack('<III', magic, tmp_val, version)
    buffer_array.append(header_buffer)

    file_count_buffer = struct.pack('<I', len(file_list))
    buffer_array.append(file_count_buffer)

    count = 0
    total = len(file_list)
    for file_path in file_list:
        file_name = os.path.relpath(file_path, input_dir)
        file_name = file_name.replace(os.sep, '/')

        with open(file_path, 'rb') as f:
            file_content = f.read()

        write_binary_string(buffer_array, file_name)

        content_length_buffer = struct.pack('<I', len(file_content))
        buffer_array.append(content_length_buffer)

        buffer_array.append(file_content)

        if count % 50 == 0:
            progress_callback(count,file_path,total)
        count += 1
    
    progress_callback(total,"",total)

    final_buffer = b''.join(buffer_array)

    with open(output_binary_file, 'wb') as out_file:
        out_file.write(final_buffer)

    print(f"Binary file created at: {output_binary_file}")


def create_crc_size_file(output_binary_file, output_crc_file):
    with open(output_binary_file, 'rb') as file:
        data_buffer = file.read()

    length = len(data_buffer)
    buffer_crc32 = zlib.crc32(data_buffer)

    output_content = f"{length}|{buffer_crc32}"
    with open(output_crc_file, 'w') as out_file:
        out_file.write(output_content)

    print(f"Crc32 result exported to: {output_crc_file}")



def compressFiles(input_directory,input_binary_file,output_binary_file,output_crc_file,progress_callback):
    create_binary_file(input_directory, input_binary_file, output_binary_file,progress_callback)
    create_crc_size_file(output_binary_file, output_crc_file)
